Add information about migrating from the old condition syntax (#1214)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/TypoScriptSyntax/Syntax/Conditions.rst

thereby determine, whether the TypoScript code following the condition
and ending where the next condition is found, should be parsed or not.

.. note::

>    The condition syntax got changed with 9.4, the old syntax removed in 10.

Reviewer: ```suggestion
   The condition syntax was changed with TYPO3v9, support for the old syntax was removed in TYPO3v10.
```

various records is combined.

.. index:: TypoScript; Symfony expression language
.. _typoscript-syntax-conditions-expression-language:

Custom conditions with the Symfony expression language
======================================================

Further information about how to extend TypoScript with your own custom

> conditions can be found within :ref:`sel-within-typoscript-conditions`.

Reviewer: This would lead to "... within _Within TypoScript conditions_" since the title from the link target is used, right?

Author: yeah

Reviewer: So we want to keep it that way?

Reviewer 2: Better write: :ref:`TypoScript conditions <sel-within-typoscript-conditions>`

Author: ok, did so, ok now?

- Conditions can be used outside of confinements (curly braces) only.
  However the :code:`[GLOBAL]` condition will always break a confinement if
  entered inside of one.

.. _typoscript-syntax-conditions-migration-sel:

> Migration from the pre-Symfony condition syntax

Reviewer: ```suggestion
Migration from the legacy condition syntax
```

===============================================

.. versionchanged:: 9.4

>    The old condition syntax got deprecated in TYPO3 version 9.4 and removed in 10

Reviewer: ```suggestion
   The old condition syntax was deprecated in TYPO3v9 and removed in TYPO3v10
```

>    with the introduction of this change

Reviewer: ```suggestion
   with the introduction of the Symfony Expression Language (SEL) in this change:
```

Reviewer: Notice that I'm not aware how well established the abbreviation "SEL" is ...

   :doc:`t3core:Changelog/9.4/Feature-85829-ImplementSymfonyExpressionLanguageForTypoScriptConditions`

> For information on how to migrate from the old pre-SEL style TypoScript conditions
> to the new SEL-conditions please see below:

Reviewer: ```suggestion
For information on how to migrate from legacy TypoScript conditions
to the new conditions please see below:
```

Variables and functions
-----------------------

For a detailed list of the available objects and functions refer to the

> :ref:`the TypoScript Reference <t3tsref:conditions>`.

Reviewer: ```suggestion
:ref:`TypoScript Reference <t3tsref:conditions>`.
```

Variables::

   [page["backend_layout"] == 1]

>       # Old syntax: [globalVar = TSFE:page|backend_layout = 1]

Reviewer: Here'd suggest using a `diff` code block.

Author: I think we can leave it like that, this is working TypoScript with just a comment ppl can copy it like that and use it.

   [END]

Arrays::

   [page["pid"] in [17,24]]
      # Old syntax: [globalVar = TSFE:page|pid=17, TSFE:page|pid=24]
      page.42.value = This page is a child of page 17 or page 24
   [END]

Migrating operators
-------------------

Please see a complete list of availible operators here:
`SEL syntax operators <https://symfony.com/doc/current/components/expression_language/syntax.html#comparison-operators>`__

Equality::

   [applicationContext == "Development"]
      # Old syntax: [applicationContext = Development]
      page.42.value = The application context is exactly "Development"
   [END]

Wildcards / regular expressions::

   [like(applicationContext, "Development*")]
      # Old syntax: [applicationContext = Development*]
      page.42.value = The application context starts with "Development"
   [END]
   [applicationContext matches "/^Development/"]
      # Old syntax: [applicationContext = Development*]
      page.42.value = The application context starts with "Development"
   [END]

Array operators::

   [17 in tree.rootLineIds || 24 in tree.rootLineIds]
      # Old syntax: [PIDinRootline = 17, 24]
   [END]

Migrating combined conditions
-----------------------------

> And conditions::

Reviewer: ```suggestion
Conjunctions (AND conditions)::
```

   [condition1() and condition2()]
      # Old syntax [ condition1 ][ condition2 ]
      page.42.value = Condition 1 and condition 2 met
   [END]

> Or conditions::

Reviewer: ```suggestion
Disjunctions (OR conditions)::
```

   [condition1() or condition2()]
      # used to be [ condition1 ] || [ condition2 ]
      temp.value = Condition 1 or condition 2 met
   [END]

Migrating user functions
------------------------

In order to migrate user functions for example
:code:`# Old syntax [userFunc = user_match(checkLocalIP, 192.168)]` provide
:ref:`custom variables <sel-ts-additional-variables>` or
:ref:`functions <sel-ts-additional-functions>` with the Symfony

> expression language.

Reviewer: ```suggestion
Expression Language.
```
